lib/yggdrasil/command/parser_builder.ex

```diff
@@ -0,0 +1,64 @@
+defmodule Yggdrasil.Command.ParserBuilder do
+  alias Yggdrasil.Command.ParseRule
+
+  defmacro __using__(_) do
+    quote do
+      Module.register_attribute __MODULE__, :parse_rules, accumulate: true
+
+      import Yggdrasil.Command.ParserBuilder
+
+      @before_compile unquote(__MODULE__)
+    end
+  end
+
+  defmacro parse(text, command) do
+    quote do
+      @parse_rules ParseRule.create(unquote(text), unquote(command))
+    end
+  end
+
+  defmacro __before_compile__(env) do
+    rules = env.module
+      |> Module.get_attribute(:parse_rules)
+      |> Enum.reverse
+      |> Enum.map(fn rule ->
+        match = Enum.map(rule.cmd_segs, fn(seg) ->
+          case seg do
+            {:literal, text} -> text
+            {:identifier, _var, atom} -> {atom, [], nil}
+          end
+        end)
+
+        attrs = rule.cmd_segs
+          |> Enum.filter(fn
+              ({:identifier, _, _}) -> true
+              ({:literal, _}) -> false
+            end)
+          |> Enum.map(fn({:identifier, _, atom}) ->
+            {atom, {atom, [], nil}}
+          end)
+
+        struct = {:%, [],
+                  [{:__aliases__, [alias: false], [rule.cmd_struct]},
+                  {:%{}, [], attrs}]}
+
+        quote do
+          defp do_parse_command(unquote(match)) do
+            {:ok, unquote(struct)}
+          end
+        end
+      end)
+
+    quote do
+      def parse(text) do
+        do_parse_command(String.split(text))
```

Comment:
I wonder based on my comment in the test section if we shouldn't force this to lower case or upper, same for the parse rules. So like parse "GO :exit" vs parse "go :exit" - while I guess it could be argued that if you type GO vs go you intended that, but if the user types go instead of GO they still intended to go... 

Don't know your test case with the mixed case just made me think about this.